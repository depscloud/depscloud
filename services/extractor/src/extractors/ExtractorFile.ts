import cheerio = require("cheerio");
import YAML = require("js-yaml");
import TOML = require("toml");

export default class ExtractorFile {
    private readonly body: string;

    constructor(body: string) {
        this.body = body;
    }

    public raw(): string {
        return this.body;
    }

    public json(): any {
        return JSON.parse(this.body);
    }

    public toml(): any {
        return TOML.parse(this.body);
    }

    public yaml(): any {
        return YAML.safeLoad(this.body);
    }

    public xml(): cheerio.Cheerio {
        return cheerio.load(this.body).root();
    }
}
