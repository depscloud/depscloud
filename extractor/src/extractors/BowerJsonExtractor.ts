import {Dependency, DependencyManagementFile} from "@depscloud/api/v1alpha/deps";
import Extractor from "./Extractor";
import ExtractorFile from "./ExtractorFile";
import Globals from "./Globals";
import Languages from "./Languages";
import MatchConfig from "../matcher/MatchConfig";

interface ID {
    organization: string;
    module: string;
}

function parseName(module: string): ID {
    let organization = Globals.ORGANIZATION;
    if (module.charAt(0) === "@") {
        const index = module.indexOf("/");
        organization = module.substr(1,  index - 1);
        module = module.substr(index + 1);
    }
    return { organization, module };
}

function extract(dependencyHash: any, scope: string): Dependency[] {
    return Object.keys(dependencyHash)
        .map((dependency) => {
            const { organization, module } = parseName(dependency);
            const versionConstraint = dependencyHash[dependency];

            return {
                organization,
                module,
                versionConstraint,
                scopes: [ scope ],
                name: dependency,
            };
        });
}

export default class BowerJsonExtractor implements Extractor {
    public matchConfig(): MatchConfig {
        return {
            includes: [
                "**/bower.json",
            ],
            excludes: [
                "**/public/**"
            ],
        };
    }

    public requires(): string[] {
        return [ "bower.json" ];
    }

    public async extract(_: string, files: { [p: string]: ExtractorFile }): Promise<DependencyManagementFile> {
        const {
            name,
            version,
            repository,
            dependencies,
            devDependencies,
            bundledDependencies,
        } = files["bower.json"].json();

        const { organization, module } = parseName(name);

        let allDependencies = extract((dependencies || {}), "");
        allDependencies = allDependencies.concat(extract((devDependencies || {}), "dev"));
        allDependencies = allDependencies.concat(extract((bundledDependencies || {}), "bundled"));

        let sourceUrl = repository;
        if (typeof repository === "object") {
            sourceUrl = repository.url;
        }

        return {
            language: Languages.JAVASCRIPT,
            system: "bower",
            sourceUrl,
            organization,
            module,
            version,
            dependencies: allDependencies,
            name,
        };
    }
}
