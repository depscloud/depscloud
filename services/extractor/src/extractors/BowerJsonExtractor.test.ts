import {readFile} from "fs";
import {promisify} from "util";
import BowerJsonExtractor from "./BowerJsonExtractor";
import ExtractorFile from "./ExtractorFile";

const readFileAsync = promisify(readFile);

describe("BowerJsonExtractor", () => {
    test("fullParse", async () => {
        const jsonPath = require.resolve("./testdata/bower.json");
        const buffer = await readFileAsync(jsonPath);
        const content = buffer.toString();

        const parser = new BowerJsonExtractor();

        const actual = await parser.extract("", { "bower.json": new ExtractorFile(content) });

        expect(actual).toMatchSnapshot();
        expect(JSON.stringify(actual, null, 2)).toMatchSnapshot();
    });
});
