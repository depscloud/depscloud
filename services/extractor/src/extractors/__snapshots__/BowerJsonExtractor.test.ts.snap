// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BowerJsonExtractor fullParse 1`] = `
Object {
  "dependencies": Array [
    Object {
      "name": "dependency",
      "scopes": Array [],
      "versionConstraint": "1.0.0",
    },
    Object {
      "name": "@second/Depedency",
      "scopes": Array [],
      "versionConstraint": "1.0.0",
    },
    Object {
      "name": "devDependency",
      "scopes": Array [
        "dev",
      ],
      "versionConstraint": "1.0.0",
    },
    Object {
      "name": "@annotationDevDependency",
      "scopes": Array [
        "dev",
      ],
      "versionConstraint": "1.0.0",
    },
    Object {
      "name": "@annotationAndSlash/DevDependency",
      "scopes": Array [
        "dev",
      ],
      "versionConstraint": "1.0.0",
    },
    Object {
      "name": "0",
      "scopes": Array [
        "bundled",
      ],
      "versionConstraint": "bundledDependency",
    },
  ],
  "language": "js",
  "name": "testName",
  "sourceUrl": "git@github.com:group/project.git",
  "system": "bower",
  "version": "1.2.3",
}
`;

exports[`BowerJsonExtractor fullParse 2`] = `
"{
  \\"language\\": \\"js\\",
  \\"system\\": \\"bower\\",
  \\"sourceUrl\\": \\"git@github.com:group/project.git\\",
  \\"version\\": \\"1.2.3\\",
  \\"dependencies\\": [
    {
      \\"name\\": \\"dependency\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": []
    },
    {
      \\"name\\": \\"@second/Depedency\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": []
    },
    {
      \\"name\\": \\"devDependency\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": [
        \\"dev\\"
      ]
    },
    {
      \\"name\\": \\"@annotationDevDependency\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": [
        \\"dev\\"
      ]
    },
    {
      \\"name\\": \\"@annotationAndSlash/DevDependency\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": [
        \\"dev\\"
      ]
    },
    {
      \\"name\\": \\"0\\",
      \\"versionConstraint\\": \\"bundledDependency\\",
      \\"scopes\\": [
        \\"bundled\\"
      ]
    }
  ],
  \\"name\\": \\"testName\\"
}"
`;
